src/Init/Data/BitVec/Lemmas.lean

```diff
@@ -4026,6 +4026,84 @@ theorem sdiv_self {x : BitVec w} :
       rcases x.msb with msb | msb <;> simp
     · rcases x.msb with msb | msb <;> simp [h]
 
+-- unsigned division never overflows
+theorem toNat_udiv_lt {w : Nat} {x y : BitVec w} :
+    x.toNat / y.toNat < 2 ^ w := by
+  have hy : y.toNat = 0 ∨ y.toNat = 1 ∨ 1 < y.toNat := by omega
+  rcases hy with hy|hy|hy
+  · simp [hy]; omega
+  · simp [hy]; omega
+  · rw [Nat.div_lt_iff_lt_mul (k := y.toNat) (x := x.toNat) (y := 2 ^ w) (by omega), show x.toNat = x.toNat * 1 by omega]
+    apply Nat.mul_lt_mul_of_le_of_lt (by omega) (by omega) (by omega)
+
+-- non-overflowing signed division bounds when numerator is positive, denumerator is positive
+theorem zero_le_sdiv_and_sdiv_lt_two_pow_of_two_le_of_pos {w : Nat} {x y : BitVec w} (hy : 2 ≤ y.toInt) (hx : 0 < x.toInt) :
+    0 ≤ x.toInt / y.toInt ∧ x.toInt / y.toInt < 2 ^ (w - 1) := by
+  rcases w with _|w
+  · simp [of_length_zero]
+  · have := Int.ediv_lt_self_of_two_le_of_zero_lt (x := x.toInt) (y := y.toInt) hy hx
+    have := Int.zero_le_udiv_of_two_le_zero_lt (x := x.toInt) (y := y.toInt) hy hx
+    have xle := le_two_mul_toInt (x := x); have xlt := two_mul_toInt_lt (x := x)
+    simp; omega
+
+-- non-overflowing signed division bounds when numerator is negative, denumerator is positive
+theorem neg_two_pow_le_sdiv_and_sdiv_lt_zero_of_two_le_of_neg {w : Nat} {x y : BitVec w} (hy : 2 ≤ y.toInt) (hx : x.toInt < 0) (hw : 0 < w) :
+    - 2 ^ (w - 1) ≤ x.toInt / y.toInt ∧ x.toInt / y.toInt < 0 := by
+  rcases w with _|w
+  · simp [hw]; omega
+  · have := Int.udiv_lt_of_two_le_lt_zero (x := x.toInt) (y := y.toInt) hy hx
+    have := Int.self_le_udiv_of_two_le_lt_zero (x := x.toInt) (y := y.toInt) hy hx
+    have xle := le_two_mul_toInt (x := x); have xlt := two_mul_toInt_lt (x := x)
+    simp; omega
+
+-- non-overflowing signed division bounds when numerator is positive, denumerator is negative
+theorem neg_two_pow_le_sdiv_and_sdiv_le_zero_of_le_neg_two_of_pos {w : Nat} {x y : BitVec w} (hy : y.toInt ≤ -2) (hx : 0 < x.toInt) (hw : 0 < w) :
+    - 2 ^ (w - 1) ≤ x.toInt / y.toInt ∧ x.toInt / y.toInt ≤ 0 := by
+  rcases w with _|w
+  · simp [hw]; omega
+  · have xle := le_two_mul_toInt (x := x); have xlt := two_mul_toInt_lt (x := x)
+    by_cases hy' : y.toInt = -1
+    · simp [hy']; omega
+    · have := Int.udiv_le_of_le_neg_two_lt_zero (x := x.toInt) (y := y.toInt) (by omega) hx
+      have := Int.neg_self_le_udiv_of_le_neg_two_zero_lt (x := x.toInt) (y := y.toInt) hy hx
+      simp; omega
+
+theorem toInt_sdiv_of_one {x y : BitVec 1} :
+    x.toInt / y.toInt = x &&& y := by
```

Comment:
There already exists `sdiv_eq_and`. Does this theorem add something or can we use `sdiv_eq_and`?

Reply:
indeed, the theorem is not necessary so I removed it